includes/class-wc-database-locks.php

```diff
@@ -0,0 +1,131 @@
+<?php
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly
+}
+
+/**
+ * Allows the DB to be locked to avoid concurrent events doing the same thing.
+ *
+ * @since    3.2.0
+ * @package  WooCommerce/Classes
+ * @category Class
+ * @author   Automattic
+ */
+class WC_Database_Locks {
+
+	/**
+	 * Request ID.
+	 * @var string
+	 */
+	private static $request_id = '';
+
+	/**
+	 * Get ID of this request.
+	 *
+	 * @return string
+	 */
+	public static function get_request_id() {
+		if ( ! self::$request_id ) {
+			self::$request_id = uniqid( wp_rand(), true );
+			add_action( 'shutdown', __CLASS__, 'release_all_locks' );
+		}
+		return self::$request_id;
+	}
+
+	/**
+	 * Get an existing lock.
+	 *
+	 * @param  string $lock_name
+	 * @return object|null
+	 */
+	public static function get_lock( $lock_name ) {
+		global $wpdb;
+		return $wpdb->get_row( $wpdb->prepare( "SELECT * FROM {$wpdb->prefix}woocommerce_locks WHERE name = %s;", $lock_name ) );
+	}
+
+	/**
+	 * Create a lock.
+	 *
+	 * @param  string $lock_name The name of this unique lock.
```

Comment:
We should document that there is a length limit on lock names. They would have to look at the db schema to see that right now.